spec/support/test_data_cache.rb

```diff
@@ -0,0 +1,71 @@
+# rubocop:disable Style/ClassVars
+class TestDataCache
+  #  A helper class for pre-creating test data with FactoryBot factories
+  #  and maintaining a cache of these objects to avoid the performance problems
+  #  of repeatedly asking FactoryBot to create trees of objects.
+  #
+  #  Refs: https://evilmartians.com/chronicles/testprof-2-factory-therapy-for-your-ruby-tests-rspec-minitest
+  #
+  # The test data and can be accessed as follows:
+  #
+  # TestDataCache.get(:course, :primary, :unpublished)
+  #
+  # Note that a course with the requested traits needs to be pre-defined in this class.
+  # It does not matter what order the traits are when passed as arguments.
+
+  class << self
+    @@cache_populated = false
+    @@courses = nil
+    @@course_factories = {
+        %i[primary unpublished] => -> do
+          FactoryBot.create(:course, :primary, :unpublished)
+        end,
+    }
+
+    def get(name, *traits)
+      case name
+      when :course
+        if @@cache_populated
+          @@courses[traits.sort] || raise_trait_error(:course, traits)
+        else
+          # fall back to factory (while we migrate all our tests)
+          @@course_factories[traits.sort].call
```

Comment:
What does failure look like if you try and call this with unregistered traits? Would it be easy to figure out what to do to fix it?

Reply:
I tried it, it's a nill error :cry: 

How about

```suggestion
          if @@course_factories.key?(traits.sort)
              @@course_factories[traits.sort].call
          else
              raise_model_error(name, traits)
          end
```